external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
     start_loc: usize,
     modifiers: Modifiers,
     context: &mut Context,
-) -> Result<Function, Box<Diagnostic>> {
+) -> Result<(Function, Option<Box<Diagnostic>>), Box<Diagnostic>> {
```

Comment:
This feels strange to me. Can we not just log the diagnostic if we have it? Why bother passing it back instead of just logging it?

Reply:
I guess what I'm suggesting is that the function parsing can locally handle trying to advance the parser to the next valid state, much like we did in the lexer, and hide the error handling complexity from the caller

Reply:
I was not sure what the best way was to structure this, but you are obviously right - we don't really have to pass the diagnostics on the success case when we can log them on the spot